### ISSUE TITLE
Lab 5.2 Test Case 11 Issues

### ISSUE THREAD

**Issue by hello-world-cell** [nus-cs2030/2223-s2#800]
Hello, I am facing issues with the self-check servers not being assigned server events at the right timings.
Is it nessesary to update the adding/removing of a customer to ALL self check servers? because i do not do that in my code.
My implementation creates Wait events only for self check 1. Then when creating ContinueWait events from Wait event - the self check server with the lowest available time is used. 

Not sure how to fix this, any help is appreciated.

<img width="344" alt="image" src="https://user-images.githubusercontent.com/67591893/230646483-dcc8858d-34b4-461a-be4b-3f27494e2594.png">

**Comment by Bwilldo** [comment]
hmm there a few things you might wanna check: 
1. did you update the self check out list every time 
2. did you only run the continue wait event when there is no self check out that can serve (prev i accidentally run continue wait event even after my self check out can serve)

**Comment by sharlynntan** [comment]
what do mean by "update the adding/removing of a customer to ALL self check servers"? 

from your code, it looks like customer 13 is waiting until self check out 3 is available. you should check that at every wait juncture, it is checking against the earliest self check out end time, instead of just self check out 3.

**Comment by edmundnwl** [comment]
the next wait event time for a customer waiting for a self-checkout should be the earliest timing of which any self-checkout is available

**Comment by easypzlemonsqueezy** [comment]
i faced a similar problem, i managed to resolve it by iterating through the self checkouts and finding the self checkout counter with the earliest next available time. you have to change your implementating of wait (and possibly intermediatewait) event(s) to account for the behaviour of self checkout. 

if a customer is waiting at a self checkout and intermediate waits at the same self checkout it may eventually be served at another self checkout

**Comment by hello-world-cell** [comment]
> hmm there a few things you might wanna check:
> 
> 1. did you update the self check out list every time
> 2. did you only run the continue wait event when there is no self check out that can serve (prev i accidentally run continue wait event even after my self check out can serve)

Hello could you elaborate your point 1?

**Comment by ahndrayo00** [comment]
Some things to check:

1. You can either update all SelfCheck queues if you have multiple SelfChecks (that's what I did). I believe other implementations include only 1 SelfCheck so there is only one queue, but you need to reference the index of SelfChecks properly then

2. The issue I faced was that I wasnt creating events based on the earliest available SelfCheck. So somewhere in your code, you need to iterate through all SelfChecks and ensure that you get the SelfCheck with the earliest FreeTime then create the event accordingly. Which means let's say you are going from Wait -> Waiting -> Serve, along the way you may need to change the SelfCheck allocated to this event multiple times.

All the best!!

**Comment by hello-world-cell** [comment]
> 1. did you update the self check out list every time

Hello, i tried the test case from #781 

this is my output : 
<img width="362" alt="image" src="https://user-images.githubusercontent.com/67591893/230702979-87095fd1-0cfc-4a94-b2ef-6358f3bf1e72.png">

there is an error with the 35th customer, not too sure where i should change my code 

right now
1) At arrive event i create wait events for the first self-check counter (because the first check counter is the only place customers can wait)
2) From Wait to Server (i check through the self-checks to see which one has the lowest avail time and return a serve event with that self check) 
3) From Wait to ContinueWait (i check through self-checks to see which one has the lowest avail time and return continuewait witht that server)
4) From ContinueWait to Serve, i repeat the same process 

each time customer is added or removed from the queue it is for that respective server and not all the self-check servers

**Comment by hello-world-cell** [comment]
> Some things to check:
> 
> 1. You can either update all SelfCheck queues if you have multiple SelfChecks (that's what I did). I believe other implementations include only 1 SelfCheck so there is only one queue, but you need to reference the index of SelfChecks properly then
> 2. The issue I faced was that I wasnt creating events based on the earliest available SelfCheck. So somewhere in your code, you need to iterate through all SelfChecks and ensure that you get the SelfCheck with the earliest FreeTime then create the event accordingly. Which means let's say you are going from Wait -> Waiting -> Serve, along the way you may need to change the SelfCheck allocated to this event multiple times.
> 
> All the best!!

could i ask how you went about doing point 1?
not too sure how wither adding or removing customers from the queue for all self checks would help in deciding which server to use for the next event
thanks for your suggestions!

**Comment by Bwilldo** [comment]
> 
Meaning when there is any changes made to your self check out, you will need to update. How i go about updating, is to create a new self check out list and replace the old one.

**Comment by Bwilldo** [comment]
> > 1. did you update the self check out list every time
> 
> Hello, i tried the test case from #781
> 
> this is my output : <img alt="image" width="362" src="https://user-images.githubusercontent.com/67591893/230702979-87095fd1-0cfc-4a94-b2ef-6358f3bf1e72.png">
> 
> there is an error with the 35th customer, not too sure where i should change my code
> 
> right now
> 
> 1. At arrive event i create wait events for the first self-check counter (because the first check counter is the only place customers can wait)
> 2. From Wait to Server (i check through the self-checks to see which one has the lowest avail time and return a serve event with that self check)
> 3. From Wait to ContinueWait (i check through self-checks to see which one has the lowest avail time and return continuewait witht that server)
> 4. From ContinueWait to Serve, i repeat the same process
> 
> each time customer is added or removed from the queue it is for that respective server and not all the self-check servers

2. From Wait to Server (i check through the self-checks to see which one has the lowest avail time and return a serve event with that self check)
do you wait to serve event? if it is, then i think you should loop through and see which self check out time is lower than your arrival time. then you pass that customer to that self check out server.

**Comment by Fastumkj** [comment]
facing the same issue right now :(

**Comment by hello-world-cell** [comment]
> facing the same issue right now :(

same here, still stuff, do let me know if you want to discuss

**Comment by cinobunzz** [comment]
am facing the exact same issue :'( please update if you do manage to solve it

**Comment by cinobunzz** [comment]
> i managed to solve it! here's what I did 
at wait, check if server is human 
if yes, the normal implementation from prev labs 
if not, then iterate through a list of self check out counters to check if any are available for service at the current time, if they are, return serve event 
if none are available, return wait event with the timing of the next shortest end time of all self check outs 
good luck!

**Comment by cinobunzz** [comment]
> am facing the exact same issue :'( please update if you do manage to solve it
i managed to solve it! here's what I did
at wait, check if server is human
if yes, the normal implementation from prev labs
if not, then iterate through a list of self check out counters to check if any are available for service at the current time, if they are, return serve event
if none are available, return wait event with the timing of the next shortest end time of all self check outs
good luck!

**Comment by McFishMealUpsize** [comment]
I believe that you creating Wait Events for all SelfCheckOut counters would solve your problem. You need to check which SelfCheckOut counter has the lowest available time in both your Wait and Continue Wait classes.

**Comment by ugchongg** [comment]
I think two things you have to make sure is that your wait event returns your continue wait event with the earliest counter _(in terms of available time)_ and when your continue wait event returns itself (when server is not available to serve the event's customer yet) it returns with the next earliest counter _(in terms of available time)_ . I created a comparator to compare the available time of the counters and used it to sort and get the earliest counter, hope this helps!

**Comment by Fastumkj** [comment]
> > facing the same issue right now :(
> 
> same here, still stuck, do let me know if you want to discuss

I managed to solve it through the waitEvent. as mentioned by @cinobunzz earlier, 
I quote: "at wait, check if server is human
if yes, normal implementation from prev labs"

extra info which I did not follow exactly "if not, then iterate through a list of self check out counters to check if any are available for service at the current time, if they are, return serve event
if none are available, return wait event with the timing of the shortest end time of all self check outs"

but afterwards(the "if not" part), I just return a new BufferEvent using this.getTime() instead of getting the server time. 

Hopefully it solves your problem as well!
